src/neoxp/Commands/ContractCommand.Download.cs

```diff
@@ -0,0 +1,75 @@
+using System;
+using System.ComponentModel.DataAnnotations;
+using System.IO;
+using System.Net.Http;
+using System.Threading.Tasks;
+using McMaster.Extensions.CommandLineUtils;
+using Neo;
+using Neo.IO.Json;
+using Neo.Network.RPC;
+
+namespace NeoExpress.Commands
+{
+    partial class ContractCommand
+    {
+        [Command(Name = "download", Description = "Download contract with storage from remote chain into local chain")]
+        internal class Download
+        {
+            readonly ExpressChainManagerFactory chainManagerFactory;
+            private RpcClient rpcClient = null!;
+            
+            public Download(ExpressChainManagerFactory chainManagerFactory)
+            {
+                this.chainManagerFactory = chainManagerFactory;
+            }
+            
+            [Argument(0, Description = "Contract invocation hash")]
+            [Required]
+            internal string Contract { get; init; } = string.Empty;
+            
+            [Argument(1, Description = "Source network RPC address")]
+            internal string Source { get; init; } = string.Empty;
+            
+            [Option(Description = "Path to neo-express data file")]
+            internal string Input { get; init; } = string.Empty;
+
+            internal async Task ExecuteAsync(TextWriter writer)
+            {
+                var (chainManager, _) = chainManagerFactory.LoadChain(Input);
+                var expressNode = chainManager.GetExpressNode();
+
+                this.rpcClient = new RpcClient(new Uri(Source));
+                
+                if (!UInt160.TryParse(Contract, out _))
+                {
+                    await writer.WriteLineAsync($"Invalid contract hash: {Contract} ").ConfigureAwait(false);
+                }
+                else
+                {
+                    // 1. Get ContractState
+                    var state = await this.rpcClient.GetContractStateAsync(Contract).ConfigureAwait(false);
+                    
+                    // 2. Get Full storage of the contract
+                    var storage_pairs = (JArray)await this.rpcClient.RpcSendAsync("getfullstorage", Contract).ConfigureAwait(false);
```

Comment:
I assume `getfullstorage` is a custom RPC endpoint? I'm not a fan of depending on an RPC method that isn't available as part of the base `RpcServer` module (except for ones exposed by Express itself obviously). Can we get `getfullstorage` added to `RpcServer`? 

Alternatively, we could use StateService RPC endpoints. Slightly more complex than `getfullstorage` but available by default on NGD provided RPC servers. The [`StateServiceStore`](https://github.com/ngdenterprise/neo-blockchaintoolkit-library/blob/master/src/bctklib/persistence/StateServiceStore.cs) used by `neo-trace` uses these endpoints. Would have the added benefit of being about to download contract + contract state at a specific blockchain height if desired.

Reply:
I understand your view and it's not ideal. I doubt we can get `getfullstorage` supported on the `RpcServer` because it potentially pulls a lot of storage (e.g. `NeoToken` storage is currently around 60 MB). Unless they're willing to accept it under the condition that we excluding native contracts.

Using `StateService` is not possible because it requires one to know which storage keys are used in the contract and not all contracts are open source. The work around is if monitor all contracts for `Storage.put/delete/get` syscalls in a 3rd party service, but that 3rd party service might as well run the RPC plugin that exposes `getfullstorage`. The `getfullstorage` is nothing more than this code
```C#
            byte[] prefix_key = StorageKey.CreateSearchPrefix(id, new byte[]{});

            JArray result = new JArray();
            foreach (var (key, value) in snapshot.Find(prefix_key))
            {
                JObject kv = new JObject();
                kv["k"] = Convert.ToBase64String(key.Key);
                kv["v"] = Convert.ToBase64String(value.Value);
                result.Add(kv);
            }

            return result;

```
I can give it a try to see what they think about it. I didn't do it because historically I've had much stronger use-cases for code additions and they've just been ignored.

Reply:
NeoTrace has the same requirement to download all the state for a contract w/o knowing the keys. The [`findstates`](https://docs.neo.org/docs/en-us/reference/rpc/latest-version/api/findstates.html) RPC method allows you to fetch all keys/value pairs for a given scriptHash at a given height. @ZhangTao1596 added this method for me specifically to enable NeoTrace.